hat/app/org/hatdex/hat/api/controllers/RichData.scala

```diff
@@ -70,342 +103,622 @@ class RichData @Inject() (
+    skip: Option[Int],
+    take: Option[Int])(implicit db: HATPostgresProfile.api.Database): Future[Result] = {
+    val dataEndpoint = s"$namespace/$endpoint"
+    val query =
+      Seq(EndpointQuery(dataEndpoint, None, None, None))
+    val data = dataService.propertyData(
+      query,
+      orderBy,
+      ordering.contains("descending"),
+      skip.getOrElse(0),
+      take.orElse(Some(defaultRecordLimit)))
+    data.map(d => Ok(Json.toJson(d)))
+  }
 
-    def bundleNotFound(bundleId: String) = ErrorMessage("Bundle Not Found", s"Bundle $bundleId not found")
+  def getContractData(
+    namespace: String,
+    endpoint: String,
+    orderBy: Option[String],
+    ordering: Option[String],
+    skip: Option[Int],
+    take: Option[Int]): Action[AnyContent] =
+    UserAwareAction.async { implicit request => //    SecuredAction(
+      {
+
+        val ret = for {
+          isOk <- isValidContractRequest(request.body.asJson)
+          ret <- if (isOk)
+            makeData(namespace, endpoint, orderBy, ordering, skip, take)
+          else Future.successful(BadRequest)
+        } yield ret
+
+        ret
+      }
+    }
 
-    def dataUpdateMissing(message: String) = ErrorMessage("Data Missing", s"Could not update records: $message")
-    def dataDeleteMissing(message: String) = ErrorMessage("Data Missing", s"Could not delete records: $message")
-    def dataLinkMissing(message: String) = ErrorMessage("Data Missing", s"Could not link records: $message")
+  private def bodyOptToContractDataRequest(
+    bodyOpt: Option[JsValue]): Option[ContractDataRequest] = {
+    for {
+      bodyAsJson <- bodyOpt
+      token <- refineV[NonEmpty]((bodyAsJson \ "token").as[String]).toOption
+      hatName <- refineV[NonEmpty]((bodyAsJson \ "hatName").as[String]).toOption
+      contractId <- refineV[NonEmpty]((bodyAsJson \ "contractId").as[String]).toOption
+      cdr = ContractDataRequest(
+        ShortLivedToken(token),
+        HatName(hatName),
+        ContractId(java.util.UUID.fromString(contractId)))
+    } yield cdr
+  }
 
-    def dataCombinatorNotFound(combinator: String) = ErrorMessage("Combinator Not Found", s"Combinator $combinator not found")
+  // TODO: Use KeyId
+  private def requestKeyId(
+    contractDataRequest: ContractDataRequest): Option[String] = {
+    for {
+      keyId <- ShortLivedTokenOps
+        .getKeyId(contractDataRequest.token.value)
+        .toOption
+    } yield keyId
+  }
 
-    def richDataDuplicate(error: Throwable) = ErrorMessage("Bad Request", s"Duplicate data - ${error.getMessage}")
-    def richDataError(error: Throwable) = ErrorMessage("Bad Request", s"Could not insert data - ${error.getMessage}")
-    def forbidden(error: Throwable) = ErrorMessage("Forbidden", s"Access Denied - ${error.getMessage}")
-    def forbidden(message: String) = ErrorMessage("Forbidden", s"Access Denied - ${message}")
+  def isValidContractRequest(bodyOpt: Option[JsValue]): Future[Boolean] = {
+    val ret = for {
```

Comment:
What is `ret`? Better name would be `maybeRet`.

Reply:
Suggestion

```
val maybeTokenValid:Option[Future[Boolean]] = for {
      cdr <- bodyOptToContractDataRequest(bodyOpt)
      keyId <- requestKeyId(cdr)
     verifiedToken = verifyTokenWithAdjudicator(cdr, keyId)
    } yield verifiedToken

maybeTokenValid.getOrElse(Future.successful(false))
```

reasoning:
- Option Match Some/None = Map+GetOrElse
- Map can be achieved in the for comprehension
- No need to pass a tuple around if possible (this is not a bad use case of tuples since they are not escaping the method scope)